### PR TITLE
feat: Add Parse Server option `resetPasswordSuccessOnInvalidEmail` to choose success or error response on password reset with invalid email

### PR DESCRIPTION
**Pull request by dblythy**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Allows the option for returning an error "A user with the email com does not exist." when an invalid password reset is called, instead of the current return true.

Closes: #7434

### Approach

Adds password policy option: resetPasswordSuccessOnInvalidEmail, which defaults to true. Does not change any existing functionality unless implicitly set.

### TODOs before merging

- [x] Add test cases
- [x] Add entry to changelog

### CODE REVIEW

**Comment by mtrezza** [comment]
Is this ready for review?

**Comment by dblythy** [comment]
I think so, I'm pretty sure the postgres tests are unrelated but not 100% sure

**Comment by mtrezza** [comment]
What you can alway do it close and re-open the PR, that will retrigger the test runs; if you don't see the "Re-run" button in the actions

**Comment by mtrezza** [comment]
I think [steps 4, 5](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#parse-server-configuration) are missing

File: src/Options/Definitions.js [hunk 1]
```diff
@@ -818,6 +818,13 @@ module.exports.PasswordPolicyOptions = {
       'Set the number of previous password that will not be allowed to be set as new password. If the option is not set or set to `0`, no previous passwords will be considered.<br><br>Valid values are >= `0` and <= `20`.<br>Default is `0`.',
     action: parsers.numberParser('maxPasswordHistory'),
   },
+  resetPasswordSuccessOnInvalidEmail: {
+    env: 'PARSE_SERVER_PASSWORD_POLICY_RESET_PASSWORD_SUCCESS_ON_INVALID_EMAIL',
+    help:
+      'Set to true if password resets should return success if the email is invalid<br><br>Default is `true`.',
```
Comment (mtrezza) [inline src/Options/Definitions.js]:
```
Set to `true`
```

File: src/Options/Definitions.js [hunk 2]
```diff
@@ -818,6 +818,13 @@ module.exports.PasswordPolicyOptions = {
       'Set the number of previous password that will not be allowed to be set as new password. If the option is not set or set to `0`, no previous passwords will be considered.<br><br>Valid values are >= `0` and <= `20`.<br>Default is `0`.',
     action: parsers.numberParser('maxPasswordHistory'),
   },
+  resetPasswordSuccessOnInvalidEmail: {
```
Comment (mtrezza) [inline src/Options/Definitions.js:910]: I think we mostly call this "passwordReset" instead of "resetPassword", also the pages templates are named that way.

Could you rename this `passwordResetSuccessOnInvalidEmail`?

File: spec/ValidationAndPasswordsReset.spec.js [hunk 3]
```diff
@@ -1082,4 +1082,56 @@ describe('Custom Pages, Email Verification, Password Reset', () => {
...
+        domain: 'd',
+      }),
+      passwordPolicy: {
+        resetPasswordSuccessOnInvalidEmail: false,
+      },
+    });
+    await expectAsync(Parse.User.requestPasswordReset('test@example.com')).toBeRejectedWith(
+      new Parse.Error(
+        Parse.Error.OBJECT_NOT_FOUND,
+        'A user with the email test@example.com does not exist.'
```
Comment (mtrezza) [inline spec/ValidationAndPasswordsReset.spec.js]: I don't think the email address should be contained in an error message. An email is PII.

File: spec/ValidationAndPasswordsReset.spec.js [hunk 4]
```diff
@@ -1082,4 +1082,56 @@ describe('Custom Pages, Email Verification, Password Reset', () => {
...
+      reconfigureServer({
+        appName: 'coolapp',
+        publicServerURL: 'http://localhost:1337/1',
+        emailAdapter: MockEmailAdapterWithOptions({
+          fromAddress: 'parse@example.com',
+          apiKey: 'k',
+          domain: 'd',
+        }),
+        passwordPolicy: {
+          resetPasswordSuccessOnInvalidEmail: [],
```
Comment (mtrezza) [inline spec/ValidationAndPasswordsReset.spec.js]: Could you add a few values here to test, like a string, a number, null, etc? There is existing code you can use for other params that just loops through.

File: src/Routers/UsersRouter.js [hunk 5]
```diff
@@ -384,7 +384,7 @@ export class UsersRouter extends ClassesRouter {
     }
   }
 
-  handleResetRequest(req) {
+  async handleResetRequest(req) {
```
Comment (mtrezza) [inline src/Routers/UsersRouter.js:417]: Nice promise chain cleanup!

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Comment by dblythy** [comment]
Can this be merged in to PS6 as a breaking change @mtrezza? Some refactoring will be needed

**Comment by mtrezza** [comment]
Yes, well spotted, this is a breaking change it seems, so perfectly suitable for PS6. I'll add it to the list.

File: spec/ValidationAndPasswordsReset.spec.js [hunk 6]
```diff
@@ -1082,4 +1082,34 @@ describe('Custom Pages, Email Verification, Password Reset', () => {
...
+        fromAddress: 'parse@example.com',
+        apiKey: 'k',
+        domain: 'd',
+      }),
+    });
+    const result = await Parse.User.requestPasswordReset('test@example.com');
+    expect(result).toEqual({});
+  });
+
+  it('should throw on an invalid reset password', async () => {
```
Comment (mtrezza) [inline spec/ValidationAndPasswordsReset.spec.js:1086]: What is the difference between these 2 tests? Both server configs look the same, so why does it throw in one test but pass in the other test - am I missing something?

**Comment by mtrezza** [comment]
@dblythy Is it correct that this is not a breaking change anymore? The default behavior will not change, but the developer can use a new Parse Server option to return an error if the email doesn't exist, correct? If so, I'll remove the `breaking` label here and we can also remove this from the Parse Server 6 list, as we can merge it anytime after release.

File: spec/ValidationAndPasswordsReset.spec.js (hunk 6 above)
Comment (dblythy) [reply spec/ValidationAndPasswordsReset.spec.js:1086]: I forgot to remove the previous test - it's old behaviour

**Comment by dblythy** [comment]
It's a breaking change - previously I had it as a server parameter but now it's change in default behaviour

**Comment by mtrezza** [comment]
Not sure if that has been discussed previously, but should we keep the default behavior to be obscure and return success? And allow the developer to change it to return an error. The opposite of how the PR does it now it seems. 
- No breaking change
- By default security through obscurity; the code comment also says: "Return success so that this endpoint can't be used to enumerate valid emails"

**Comment by dblythy** [comment]
The whole point of removing the obscurity was that you can check whether a user exists by calling `.signup`

https://github.com/parse-community/parse-server/security/advisories/GHSA-cfgm-jqhh-xmhr#advisory-comment-64808

**Comment by mtrezza** [comment]
I mean because we'd set a lower security option as default. It's technically true that the sign-up endpoint may also reveal whether an account exists, but we're making it easier for an attacker who may not think of that.

**Comment by mtrezza** [comment]
OK, so let's not make this a breaking change by setting the default to always return true, if we can't find an argument why the default setting should be less secure (obscure).

**Review by mtrezza — CHANGES_REQUESTED** [review]
The main reason for removing the obscurity is to give developers the ability to implement [their own flow](https://github.com/parse-community/parse-server/security/advisories/GHSA-cfgm-jqhh-xmhr#advisory-comment-64808).

I think we may need to add a Parse Server option to choose whether to obscure or not, because of client app migration. Imagine a client app that always expects a 200 response and displays the usual message "if we find an account for that email address, we will send a password reset email". That logic will break, because if the server returns a 400 response, the client may not have a logic implemented for that or just display a common error message. We would force a business to update all their mobile clients before they upgrade to Parse Server 6. That's not what we want as it decreases adoption.

Once we have the option, I still think we should choose the default to not be transparent, i.e. keep the status quo. Even if it doesn't prevent finding out about email addresses, it makes it harder (1 vulnerable endpoint less). Imagine an app that doesn't expose the sign-up endpoint, in which case it would make a lot of sense to keep the obscurity.

File: src/Options/index.js [hunk 7]
```diff
@@ -525,6 +525,11 @@ export interface PasswordPolicyOptions {
   Default is `false`.
   :DEFAULT: false */
   resetTokenReuseIfValid: ?boolean;
+  /* Set to true if password resets should return success if the email is invalid
```
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Set to `true` if a request to reset the password should return a success response even if the provided email address is invalid, or `false` if the request should return an error response if the email address is invalid.
```

**Review by mtrezza — DISMISSED** [review]
Looks good!

File: src/Options/Definitions.js [hunk 8]
```diff
@@ -907,6 +907,13 @@ module.exports.PasswordPolicyOptions = {
...
+    env: 'PARSE_SERVER_PASSWORD_POLICY_RESET_PASSWORD_SUCCESS_ON_INVALID_EMAIL',
+    help:
+      'Set to true if password resets should return success if the email is invalid<br><br>Default is `true`.',
```
Comment (mtrezza) [inline src/Options/Definitions.js]:
```suggestion
      'Set to `true` if a request to reset the password should return a success response even if the provided email address is invalid, or `false` if the request should return an error response if the email address is invalid.<br><br>Default is `true`.',
```

File: src/Options/docs.js [hunk 9]
```diff
@@ -207,6 +207,7 @@
  * @property {Boolean} doNotAllowUsername Set to `true` to disallow the username as part of the password.<br><br>Default is `false`.
  * @property {Number} maxPasswordAge Set the number of days after which a password expires. Login attempts fail if the user does not reset the password before expiration.
  * @property {Number} maxPasswordHistory Set the number of previous password that will not be allowed to be set as new password. If the option is not set or set to `0`, no previous passwords will be considered.<br><br>Valid values are >= `0` and <= `20`.<br>Default is `0`.
+ * @property {Boolean} resetPasswordSuccessOnInvalidEmail Set to true if password resets should return success if the email is invalid<br><br>Default is `true`.
```
Comment (mtrezza) [inline src/Options/docs.js]:
```suggestion
 * @property {Boolean} resetPasswordSuccessOnInvalidEmail Set to `true` if a request to reset the password should return a success response even if the provided email address is invalid, or `false` if the request should return an error response if the email address is invalid.<br><br>Default is `true`.
```

**Review by mtrezza — APPROVED** [review]

**Comment by mtrezza** [comment]
I've had another look at the options. Should this go under `PasswordPolicyOptions` since we also have things like `resetTokenValidityDuration` in there?

So this could be named `passwordPolicy.resetResponseSuccessOnInvalidEmail`. On the other hand it may not be strictly understood as a "password policy".

Anyway, let's just merge and hopefully someday restructure all options in a more meaningful way.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.34](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.34)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/6.1.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0](https://github.com/parse-community/parse-server/releases/tag/6.1.0)
